src/locale.js

```diff
@@ -357,9 +357,14 @@ function parseYear(d, string, i) {
 }
 
 function parseZone(d, string, i) {
-  return /^[+-]\d{4}$/.test(string = string.slice(i, i + 5))
-      ? (d.Z = -string, i + 5) // sign differs from getTimezoneOffset!
-      : -1;
+  var n = /^(Z)|([+-]\d\d)(?:\:?(\d\d))?/.exec(string.slice(i));
+  if (n) {
+    d.Z = n[1]
+        ? 0
+        : -(n[2] + (n[3] || "00")); // sign differs from getTimezoneOffset!
```

Comment:
Even though it’s slightly more code, it seems better to convert to a number as soon as possible and then add the implicit zeroes by multiplying by 100.

```js
d.Z = n[1] ? 0 // Zulu time
    : n[3] ? -(n[2] + n[3]) // sign differs from getTimezoneOffset!
    : -n[2] * 100;
```